### PR TITLE
Allow job_request_items in ArmadaOperator to be templated

### LINKED ISSUE #2482: Add templating support in Airflow ArmadaOperator
Airflow provides a “templating” (jinja2) features for Operators fields. 
They are in practice the only way to communicate values from a task to another (including XCOM).

For example (just including templated fields):

```
    kpo_read_xcom_env = KubernetesPodOperator(
        arguments=["""echo {{ ti.xcom_pull(task_ids='Start', key='return_value')['myvalue'] }} """],

     	env_vars={
            "XCOM_VALUE": """{{ ti.xcom_pull(task_ids='Start', key='return_value')['myvalue'] }}"""
        },
    )
```

Normally for simple fields such as strings or collection of strings you get templating for free just setting the template_fields class variable.

```
template_fields: Sequence[str] = (
         'arguments', 'env_vars'
    )
```

For more complex fields some method overriding. An example of this is in KPO https://github.com/apache/airflow/blob/3363004450355582712272924fac551dc1f7bd56/airflow/providers/cncf/kubernetes/operators/pod.py#L395-L423

Tried to do a similar thing for ArmadaOperator, something like:

```
class ArmadaOperator2(ArmadaOperator):
    template_fields: Sequence[str] = (
         'job_request_items', 'name'
    )

    def _render_nested_template_fields(
        self,
        content: Any,
        context: Context,
        jinja_env: jinja2.Environment,
        seen_oids: set,
    ) -> None:
        if id(content) not in seen_oids:
            template_fields: tuple | None

            if isinstance(content, submit_pb2.JobSubmitRequestItem):
                template_fields = ("pod_spec",)
            elif isinstance(content, core_v1.Container):
                template_fields = ("args",)
            else:
                template_fields = None

            if template_fields:
                seen_oids.add(id(content))
                self._do_render_template_fields(content, template_fields, context, jinja_env, seen_oids)
                return

        super()._render_nested_template_fields(content, context, jinja_env, seen_oids)
```

But this does not work 

```
Traceback (most recent call last):
  File "/airflow/armada-test.py", line 194, in <module>
    dag.test()
  File "/airflow/venv/lib64/python3.9/site-packages/airflow/utils/session.py", line 75, in wrapper
    return func(*args, session=session, **kwargs)
  File "/airflow/venv/lib64/python3.9/site-packages/airflow/models/dag.py", line 2552, in test
    _run_task(ti, session=session)
  File "/airflow/venv/lib64/python3.9/site-packages/airflow/models/dag.py", line 3655, in _run_task
    ti._run_raw_task(session=session)
  File "/airflow/venv/lib64/python3.9/site-packages/airflow/utils/session.py", line 72, in wrapper
    return func(*args, **kwargs)
  File "/airflow/venv/lib64/python3.9/site-packages/airflow/models/taskinstance.py", line 1383, in _run_raw_task
    self._execute_task_with_callbacks(context, test_mode)
  File "/airflow/venv/lib64/python3.9/site-packages/airflow/models/taskinstance.py", line 1502, in _execute_task_with_callbacks
    task_orig = self.render_templates(context=context)
  File "/airflow/venv/lib64/python3.9/site-packages/airflow/models/taskinstance.py", line 2123, in render_templates
    original_task.render_template_fields(context)
  File "/airflow/venv/lib64/python3.9/site-packages/airflow/models/baseoperator.py", line 1179, in render_template_fields
    self._do_render_template_fields(self, self.template_fields, context, jinja_env, set())
  File "/airflow/venv/lib64/python3.9/site-packages/airflow/utils/session.py", line 75, in wrapper
    return func(*args, session=session, **kwargs)
  File "/airflow/venv/lib64/python3.9/site-packages/airflow/models/abstractoperator.py", line 603, in _do_render_template_fields
    rendered_content = self.render_template(
  File "/airflow/venv/lib64/python3.9/site-packages/airflow/models/abstractoperator.py", line 676, in render_template
    return [self.render_template(element, context, jinja_env, oids) for element in value]
  File "/airflow/venv/lib64/python3.9/site-packages/airflow/models/abstractoperator.py", line 676, in <listcomp>
    return [self.render_template(element, context, jinja_env, oids) for element in value]
  File "/airflow/venv/lib64/python3.9/site-packages/airflow/models/abstractoperator.py", line 683, in render_template
    self._render_nested_template_fields(value, context, jinja_env, oids)
  File "/airflow/armada-test.py", line 146, in _render_nested_template_fields
    self._do_render_template_fields(content, template_fields, context, jinja_env, seen_oids)
  File "/airflow/venv/lib64/python3.9/site-packages/airflow/utils/session.py", line 75, in wrapper
    return func(*args, session=session, **kwargs)
  File "/airflow/venv/lib64/python3.9/site-packages/airflow/models/abstractoperator.py", line 619, in _do_render_template_fields
    setattr(parent, attr_name, rendered_content)
AttributeError: Assignment not allowed to field "pod_spec" in protocol message object.
```

At the minimum it is desirable that the following parameters in podpec.containers[] are templatable:

command
args

Airflow provides a “templating” (jinja2) features for Operators fields. 
They are in practice the only way to communicate values from a task to another (including XCOM).

For example (just including templated fields):

```
    kpo_read_xcom_env = KubernetesPodOperator(
        arguments=["""echo {{ ti.xcom_pull(task_ids='Start', key='return_value')['myvalue'] }} """],

     	env_vars={
            "XCOM_VALUE": """{{ ti.xcom_pull(task_ids='Start', key='return_value')['myvalue'] }}"""
        },
    )
```

Normally for simple fields such as strings or collection of strings you get templating for free just setting the template_fields class variable.

```
template_fields: Sequence[str] = (
         'arguments', 'env_vars'
    )
```

For more complex fields some method overriding is needed. An example of this is in KPO https://github.com/apache/airflow/blob/3363004450355582712272924fac551dc1f7bd56/airflow/providers/cncf/kubernetes/operators/pod.py#L395-L423

Tried to do a similar thing for ArmadaOperator, something like:

```
class ArmadaOperator2(ArmadaOperator):
    template_fields: Sequence[str] = (
         'job_request_items', 'name'
    )

    def _render_nested_template_fields(
        self,
        content: Any,
        context: Context,
        jinja_env: jinja2.Environment,
        seen_oids: set,
    ) -> None:
        if id(content) not in seen_oids:
            template_fields: tuple | None

            if isinstance(content, submit_pb2.JobSubmitRequestItem):
                template_fields = ("pod_spec",)
            elif isinstance(content, core_v1.Container):
                template_fields = ("args",)
            else:
                template_fields = None

            if template_fields:
                seen_oids.add(id(content))
                self._do_render_template_fields(content, template_fields, context, jinja_env, seen_oids)
                return

        super()._render_nested_template_fields(content, context, jinja_env, seen_oids)
```

But this does not work 

```
Traceback (most recent call last):
  File "/airflow/armada-test.py", line 194, in <module>
    dag.test()
  File "/airflow/venv/lib64/python3.9/site-packages/airflow/utils/session.py", line 75, in wrapper
    return func(*args, session=session, **kwargs)
  File "/airflow/venv/lib64/python3.9/site-packages/airflow/models/dag.py", line 2552, in test
    _run_task(ti, session=session)
  File "/airflow/venv/lib64/python3.9/site-packages/airflow/models/dag.py", line 3655, in _run_task
    ti._run_raw_task(session=session)
  File "/airflow/venv/lib64/python3.9/site-packages/airflow/utils/session.py", line 72, in wrapper
    return func(*args, **kwargs)
  File "/airflow/venv/lib64/python3.9/site-packages/airflow/models/taskinstance.py", line 1383, in _run_raw_task
    self._execute_task_with_callbacks(context, test_mode)
  File "/airflow/venv/lib64/python3.9/site-packages/airflow/models/taskinstance.py", line 1502, in _execute_task_with_callbacks
    task_orig = self.render_templates(context=context)
  File "/airflow/venv/lib64/python3.9/site-packages/airflow/models/taskinstance.py", line 2123, in render_templates
    original_task.render_template_fields(context)
  File "/airflow/venv/lib64/python3.9/site-packages/airflow/models/baseoperator.py", line 1179, in render_template_fields
    self._do_render_template_fields(self, self.template_fields, context, jinja_env, set())
  File "/airflow/venv/lib64/python3.9/site-packages/airflow/utils/session.py", line 75, in wrapper
    return func(*args, session=session, **kwargs)
  File "/airflow/venv/lib64/python3.9/site-packages/airflow/models/abstractoperator.py", line 603, in _do_render_template_fields
    rendered_content = self.render_template(
  File "/airflow/venv/lib64/python3.9/site-packages/airflow/models/abstractoperator.py", line 676, in render_template
    return [self.render_template(element, context, jinja_env, oids) for element in value]
  File "/airflow/venv/lib64/python3.9/site-packages/airflow/models/abstractoperator.py", line 676, in <listcomp>
    return [self.render_template(element, context, jinja_env, oids) for element in value]
  File "/airflow/venv/lib64/python3.9/site-packages/airflow/models/abstractoperator.py", line 683, in render_template
    self._render_nested_template_fields(value, context, jinja_env, oids)
  File "/airflow/armada-test.py", line 146, in _render_nested_template_fields
    self._do_render_template_fields(content, template_fields, context, jinja_env, seen_oids)
  File "/airflow/venv/lib64/python3.9/site-packages/airflow/utils/session.py", line 75, in wrapper
    return func(*args, session=session, **kwargs)
  File "/airflow/venv/lib64/python3.9/site-packages/airflow/models/abstractoperator.py", line 619, in _do_render_template_fields
    setattr(parent, attr_name, rendered_content)
AttributeError: Assignment not allowed to field "pod_spec" in protocol message object.
```

At the minimum it is desirable that the following parameters in podspec.containers[] are templatable:

command
args

I think that this feature will require some deep method overriding.

┆Issue is synchronized with this [Jira Task](https://gr-oss.atlassian.net/browse/BATCH-296) by [Unito](https://www.unito.io)

### PR DESCRIPTION
**Pull request by gr-robinlevins**
For Issue 2482:
https://github.com/armadaproject/armada/issues/2482

#### What type of PR is this?
This adds templating to the Airflow ArmadaOperator, to enable job_submit_items to be templated natively in Airflow.

#### What this PR does / why we need it:
See above

#### Which issue(s) this PR fixes:
Fixes #2482

#### Special notes for your reviewer:
Let me know if you have any questions. Thanks!

┆Issue is synchronized with this [Jira Task](https://gr-oss.atlassian.net/browse/BATCH-333) by [Unito](https://www.unito.io)

### CODE REVIEW

File: third_party/airflow/armada/operators/armada.py [hunk 1]
```diff
@@ -131,3 +139,12 @@ def _get_lookout_url(self, job_id: str) -> str:
         if self.lookout_url_template is None:
             return ""
         return self.lookout_url_template.replace("<job_id>", job_id)
+
+    def render_template_fields(
```
Comment (ClifHouck) [inline third_party/airflow/armada/operators/armada.py:143]: Can you add this method to `ArmadaDeferrableOperator` as well? It'd be nice to keep feature parity between the two.

**Review by ClifHouck — CHANGES_REQUESTED** [review]
Could you add a couple of unit tests to confirm templating works as expected on `job_request_items`?

File: third_party/airflow/armada/operators/armada.py (hunk 1 above)
Comment (gr-robinlevins) [reply third_party/airflow/armada/operators/armada.py:143]: Will do! I'll add that and the unit tests.

**Comment by gr-robinlevins** [comment]
PR updated, please give it a look! Thanks!

**Comment by kannon92** [comment]
> PR updated, please give it a look! Thanks!

I think you gotta apply python formatting changes (`tox -e format-code`).  Or something like that.

**Comment by gr-robinlevins** [comment]
Formatting changes applied!

**Comment by gr-robinlevins** [comment]
I admit I'm struggling to understand why the automated tooling has failed this pass. Looking at this:
https://github.com/armadaproject/armada/actions/runs/5248672610/jobs/9484210027#step:6:280
It looks like it's complaining about the generated docs looking different from the actual docs, is that correct? But the diff provided looks identical. I think I'm misreading or misunderstanding something. What am I missing here? Thanks!

**Comment by gr-robinlevins** [comment]
Got it - will do!

**Review by ClifHouck — APPROVED** [review]
LGTM, thanks for this!

**Comment by gr-robinlevins** [comment]
Alright, hopefully this docs update has done the trick!

**Review by kannon92 — APPROVED** [review]

**Comment by kannon92** [comment]
> It's not clear to me why the integration test is failing, and it's not one I can run locally. But it looks like a network error, and [I can see someone else's Python integration test failed similarly](https://github.com/armadaproject/armada/actions/runs/5281816301/jobs/9555897394), so maybe it's a pipeline issue?

We are investigating that now but it isn’t related to your code. Merged! Thank you for your contribution!

**Pull request merged**